pymbolic/interop/common.py

```diff
@@ -61,167 +66,201 @@ def not_supported(self, expr):
 
 # {{{ sympy like -> pymbolic
 
-class SympyLikeToPymbolicMapper(SympyLikeMapperBase):
+class SympyLikeToPymbolicMapper(SympyLikeMapperBase[Expression, []]):
+
+    def rec_arith(self, expr: object) -> ArithmeticExpression:
+        result = self.rec(expr)
+        assert prim.is_arithmetic_expression(result)
+
+        return result
 
     # {{{ utils
 
-    def to_float(self, expr):
+    def to_float(self, expr: optype.CanFloat) -> Expression:
         return float(expr)
 
-    def function_name(self, expr):
+    def function_name(self, expr: object) -> str:
         # Given a symbolic function application f(x), return the name of f as a
         # string
-        raise NotImplementedError()
+        raise NotImplementedError(f"'function_name' for type {type(expr)}")
 
     # }}}
 
-    def map_Symbol(self, expr):
+    def map_Symbol(self, expr: sp.Symbol) -> Expression:
```

Comment:
This stuff is lying because it can take `sympy.Symbol` or `symengine.Symbol`. Not quite sure how to make this work nicely?